File: ruby/lib/mable_bank/application/processing_report.rb
# frozen_string_literal: true

module MableBank
  module Application
    class ProcessingReport
      attr_reader :recorder, :dry_run, :transfer_journal, :run_id

      def initialize(recorder:, account_book:, dry_run: false, transfer_journal: nil, run_id: nil)
        @recorder = recorder
        @account_book = account_book
        @dry_run = dry_run
        @transfer_journal = transfer_journal
        @run_id = run_id
      end

      def final_balances
        @account_book.final_balances
      end

      def dry_run?
        @dry_run
      end

      def processed_count
        @recorder.processed_count
      end

      def succeeded_count
        @recorder.succeeded_count
      end

      def failed_count
        @recorder.failed_count
      end

      def failed_results
        @recorder.failed_results
      end

      def skipped_count
        @recorder.skipped_count
      end

      def journal_entries
        @transfer_journal&.entries || []
      end
    end
  end
end

File: dotnet/src/MableBank.Application/ProcessingReport.cs
using MableBank.Domain;

namespace MableBank.Application;

public sealed class ProcessingReport
{
    public ProcessingReport(
        TransferResultRecorder recorder,
        CompanyAccountBook accountBook,
        bool dryRun,
        ITransferJournal? transferJournal = null,
        Guid? runId = null)
    {
        Recorder = recorder;
        AccountBook = accountBook;
        DryRun = dryRun;
        TransferJournal = transferJournal;
        RunId = runId;
    }

    public TransferResultRecorder Recorder { get; }

    public CompanyAccountBook AccountBook { get; }

    public bool DryRun { get; }

    public ITransferJournal? TransferJournal { get; }

    public Guid? RunId { get; }

    public IReadOnlyList<AccountBalance> FinalBalances => AccountBook.FinalBalances();

    public int ProcessedCount => Recorder.ProcessedCount;

    public int SucceededCount => Recorder.SucceededCount;

    public int FailedCount => Recorder.FailedCount;

    public IReadOnlyList<TransferResult> FailedResults => Recorder.FailedResults;

    public int SkippedCount => Recorder.SkippedCount;

    public IReadOnlyList<TransferJournalEntry> JournalEntries =>
        TransferJournal?.Entries ?? [];
}
